SU2_CFD/src/output/filewriter/CCGNSFileWriter.cpp

```diff
@@ -0,0 +1,310 @@
+  cumulative = 0;
+
+  /*--- If surface file cell dimension is decreased. ---*/
+  nCell = nDim - isSurface;
+
+  if (rank == MASTER_NODE) {
+    /*--- Remove the previous file if present. ---*/
+    remove(fileName.c_str());
+
+    /*--- Create CGNS file and open in write mode. ---*/
+    CallCGNS(cg_open(fileName.c_str(), CG_MODE_WRITE, &cgnsFileID));
+
+    /*--- Create Base. ---*/
+    CallCGNS(cg_base_write(cgnsFileID, "Base", nCell, nDim, &cgnsBase));
+
+    /*--- Create Zone. ---*/
+    vector<cgsize_t> zoneData(3);
+
+    zoneData[0] = GlobalPoint;
+    zoneData[1] = GlobalElem;
+    zoneData[2] = 0;
+
+    CallCGNS(cg_zone_write(cgnsFileID, cgnsBase, "Zone", zoneData.data(), Unstructured, &cgnsZone));
+  }
+}
+
+void CCGNSFileWriter::WriteCoordinate(int CoordinateNumber, const string& CoordinateName) {
+  cgsize_t nodeBegin;
+  cgsize_t nodeEnd;
+  vector<float> sendBufferFloat, recvBufferFloat;
+
+  /*--- Create send buffer. ---*/
+  sendBufferFloat.resize(nLocalPoints);
+
+  unsigned short iCoord = CoordinateNumber - 1;
+  for (int iPoint = 0; iPoint < nLocalPoints; iPoint++) {
+    sendBufferFloat[iPoint] = static_cast<float>(dataSorter->GetData(iCoord, iPoint));
+  }
+
+#ifdef HAVE_MPI
+  SU2_MPI::Request sendReq;
+  if (rank != MASTER_NODE)
+    SU2_MPI::Isend(sendBufferFloat.data(), nLocalPoints, MPI_FLOAT, MASTER_NODE, 0, SU2_MPI::GetComm(), &sendReq);
+#endif
+
+  /*--- Coordinate vector is written in blocks, one for each process. ---*/
+  if (rank == MASTER_NODE) {
+    nodeBegin = 1;
+    nodeEnd = static_cast<cgsize_t>(nLocalPoints);
+    CallCGNS(cg_coord_partial_write(cgnsFileID, cgnsBase, cgnsZone, RealSingle, CoordinateName.c_str(), &nodeBegin,
+                                    &nodeEnd, sendBufferFloat.data(), &CoordinateNumber));
+  }
+
+#ifdef HAVE_MPI
+  if (rank == MASTER_NODE) {
+    for (int i = 1; i < size; ++i) {
+      nodeBegin = static_cast<cgsize_t>(dataSorter->GetnPointCumulative(i) + 1);
+      nodeEnd = static_cast<cgsize_t>(dataSorter->GetnPointCumulative(i + 1));
+
+      int recvSize = static_cast<int>(nodeEnd - nodeBegin + 1);
+      recvBufferFloat.resize(recvSize);
+
+      SU2_MPI::Recv(recvBufferFloat.data(), recvSize, MPI_FLOAT, i, 0, SU2_MPI::GetComm(), MPI_STATUS_IGNORE);
+      if (recvSize <= 0) continue;
+      CallCGNS(cg_coord_partial_write(cgnsFileID, cgnsBase, cgnsZone, RealSingle, CoordinateName.c_str(), &nodeBegin,
+                                      &nodeEnd, recvBufferFloat.data(), &CoordinateNumber));
+    }
+  }
+
+  if (rank != MASTER_NODE) SU2_MPI::Wait(&sendReq, MPI_STATUS_IGNORE);
+#endif
+}
+
+void CCGNSFileWriter::WriteConnectivity(GEO_TYPE type, const string& SectionName) {
+  cgsize_t firstElem, endElem;
+
+  unsigned long iElem, iPoint;
+  unsigned long nLocalElem, nTotElem;
+  unsigned short nPointsElem;
+  int cgnsSection;
+
+  ElementType_t elementType;
+
+#ifdef HAVE_MPI
+  SU2_MPI::Request sendReq;
+#endif
```

Comment:
Try to avoid the multiple `ifdef HAVE_MPI`, the MPI wrapper can be called by sequential code.
It is safer to have just one configuration to test.